tools/jar2shrinkwrap/src/main/java/jakartatck/jar2shrinkwrap/WarFileProcessor.java

```diff
@@ -55,48 +78,72 @@ public void saveOutput(final File fileInputArchive) {
-            printWriter.println("\"" + testclient + ".war\")");
-            for (String name : libraries) {
-                printWriter.print(".addAsLibrary(");
-                printWriter.print(name);
-                printWriter.println(")");
+            saveOutput(fileWriter, true);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void saveOutput(Writer writer, boolean includeImports) {
+        String testclient = "Client";
+        String indent = "\t";
+        try (PrintWriter printWriter = new PrintWriter(writer)) {
+            if(includeImports) {
+                printWriter.println("import org.jboss.arquillian.container.test.api.Deployment;");
+                printWriter.println("import org.jboss.shrinkwrap.api.ShrinkWrap;");
+                printWriter.println("import org.jboss.shrinkwrap.api.spec.WebArchive;\n");
             }
+
+            printWriter.println(indent+"@Deployment(testable = false)");
+            printWriter.println(indent+"public static WebArchive getTestArchive() throws Exception {");
+            // The libary jars
+            printWriter.println(indent.repeat(2)+"List<File> libraryFiles = new ArrayList<>();");
+            printWriter.println(indent.repeat(2)+"for (String jarName : war.getLibraries()) {");
+            printWriter.println(indent.repeat(3)+"File jarFile = new File(war.getLibDir(), jarName);");
+            printWriter.println(indent.repeat(3)+"libraryFiles.add(jarFile);");
+            printWriter.println(indent.repeat(2)+"}");
+            printWriter.println(indent.repeat(2)+"List<JavaArchive> warJars = libraryFiles.stream()\n"+
+                    indent.repeat(2)+".map(file -> ShrinkWrap.createFromZipFile(JavaArchive.class, file))\n"+
+                    indent.repeat(2)+".collect(Collectors.toList());");
+
+            // Start war creation
+            printWriter.print(indent.repeat(2)+"return ShrinkWrap.create(WebArchive.class, ");
+            printWriter.println("\"" + testclient + ".war\")");
+
+            printWriter.println(indent.repeat(3)+".addAsLibraries(warJars)");
+
             for (String name : classes) {
-                printWriter.print(".addClass(");
+                printWriter.print(indent.repeat(3)+".addClass(");
                 printWriter.print(name);
-                printWriter.println(")");
+                printWriter.println(".class)");
             }
             for (String name : webinf) {
-                printWriter.print(".addAsWebInfResource(");
+                printWriter.print(indent.repeat(3)+".addAsWebInfResource(\"");
                 printWriter.print(name);
-                printWriter.println(")");
+                printWriter.println("\")");
             }
+            // I don't think this is valid in general as a custom manifest would not be added to a deployment
+            /*
```

Comment:
Hmm, https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0.html#web-application-archive-file mentions the following about  war archives including a META-INF folder:

> Web applications can be packaged and signed into a Web ARchive format (WAR) file using the standard Java archive tools. For example, an application for issue tracking might be distributed in an archive file called issuetrack.war.
> 
> When packaged into such a form, a META-INF directory will be present which contains information useful to Java archive tools. This directory must not be directly served as content by the container in response to a web client’s request, though its contents are visible to servlet code via the getResource and getResourceAsStream calls on the ServletContext. Also, any requests to access the resources in META-INF directory must be returned with a SC_NOT_FOUND (404) response.

Reply:
Right, it might be there due to the tools, but it is not part of the application. We should not be adding it in the deployment method.